core/src/main/java/jenkins/plugins/DetachedPluginsManager.java

```diff
@@ -0,0 +1,227 @@
+package jenkins.plugins;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import hudson.ClassicPluginStrategy;
+import hudson.PluginWrapper;
+import hudson.util.VersionNumber;
+import jenkins.util.java.JavaUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Dedicated class to handle the logic related to so-called <em>detached plugins</em>.
+ *
+ * <p>Originally, some features were directly in Jenkins core. Over time, more and more features got extracted in dedicated plugins.
+ * Issue is: many plugins had started depending on these features, that now were not in the core anymore. FIXME
+ * So the chosen design strategy has been that the jenkins.war would embed these plugins, and automatically install them and mark them as optional dependencies.
+ * This way, older plugins would keep working without having to be modified.</p>
+ * <p>
+ * This code was originally moved from {@link ClassicPluginStrategy}.
+ *
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class DetachedPluginsManager {
```

Comment:
The name looks confusing since it does not implement "PluginManager". Add "Helper" to the name?

Reply:
I see the point. It's more an old habit of calling things that "manage" something a Thing__Manager__. Or Handler. _Helper_ I'm _slightly_ uncomfortable with, but will think about it.